File: budget_tracker.js
// budget_tracker.js
const fs = require('fs');
const readline = require('readline');

const DATA_FILE = 'budget.json';
let transactions = [];

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function ask(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function load() {
    try {
        if (fs.existsSync(DATA_FILE)) {
            const data = fs.readFileSync(DATA_FILE, 'utf8');
            transactions = JSON.parse(data);
        } else {
            transactions = [];
        }
    } catch (e) {
        transactions = [];
    }
}

function save() {
    fs.writeFileSync(DATA_FILE, JSON.stringify(transactions, null, 2));
}

function addTransaction(type, category, amount, dateStr) {
    if (!dateStr || dateStr.trim() === '') {
        dateStr = new Date().toISOString().slice(0, 10);
    } else {
        // validate date format
        if (!/^\d{4}-\d{2}-\d{2}$/.test(dateStr)) {
            console.log('Invalid date format. Use YYYY-MM-DD.');
            return false;
        }
    }
    transactions.push({
        type: type,
        category: category,
        amount: parseFloat(amount),
        date: dateStr
    });
    save();
    return true;
}

function getBalance() {
    let income = 0, expense = 0;
    for (const t of transactions) {
        if (t.type === 'income') income += t.amount;
        else if (t.type === 'expense') expense += t.amount;
    }
    return { income, expense, balance: income - expense };
}

function categoryStats(type) {
    const stats = {};
    for (const t of transactions) {
        if (t.type === type) {
            if (!stats[t.category]) stats[t.category] = 0;
            stats[t.category] += t.amount;
        }
    }
    return stats;
}

function printChart(type, width = 40) {
    const stats = categoryStats(type);
    const keys = Object.keys(stats);
    if (keys.length === 0) {
        console.log('No data to chart.');
        return;
    }
    let total = 0, maxVal = 0;
    for (const k of keys) {
        total += stats[k];
        if (stats[k] > maxVal) maxVal = stats[k];
    }
    console.log(`\n${type.charAt(0).toUpperCase() + type.slice(1)} chart (by category):`);
    for (const cat of keys.sort((a, b) => stats[b] - stats[a])) {
        const barLen = Math.round((stats[cat] / maxVal) * width);
        const bar = '█'.repeat(barLen);
        console.log(`${cat.padEnd(12)} ${bar} ${stats[cat].toFixed(2)}`);
    }
    console.log(`Total: ${total.toFixed(2)}`);
}

async function main() {
    load();
    console.log('=== Budget Tracker ===');
    while (true) {
        console.log('\n1. Add transaction');
        console.log('2. Show balance');
        console.log('3. Show category stats');
        console.log('4. Show expense chart');
        console.log('5. Save & exit');
        const choice = await ask('Choose: ');
        switch (choice.trim()) {
            case '1': {
                const type = (await ask('Type (income/expense): ')).trim().toLowerCase();
                if (!['income', 'expense'].includes(type)) {
                    console.log('Invalid type.');
                    break;
                }
                const category = (await ask('Category: ')).trim();
                const amountStr = await ask('Amount: ');
                const amount = parseFloat(amountStr);
                if (isNaN(amount) || amount < 0) {
                    console.log('Invalid amount.');
                    break;
                }
                const dateStr = (await ask('Date (YYYY-MM-DD, leave blank for today): ')).trim();
                if (addTransaction(type, category, amount, dateStr)) {
                    console.log('Transaction added.');
                }
                break;
            }
            case '2': {
                const { income, expense, balance } = getBalance();
                console.log(`\nIncome: ${income.toFixed(2)}`);
                console.log(`Expenses: ${expense.toFixed(2)}`);
                console.log(`Balance: ${balance.toFixed(2)}`);
                break;
            }
            case '3': {
                const stats = categoryStats('expense');
                const keys = Object.keys(stats);
                if (keys.length === 0) {
                    console.log('No expenses recorded.');
                } else {
                    let total = 0;
                    for (const k of keys) total += stats[k];
                    console.log('\nExpense categories:');
                    for (const cat of keys.sort((a, b) => stats[b] - stats[a])) {
                        const pct = (stats[cat] / total * 100);
                        console.log(`  ${cat.padEnd(12)} ${stats[cat].toFixed(2).padStart(10)}  (${pct.toFixed(1)}%)`);
                    }
                }
                break;
            }
            case '4':
                printChart('expense');
                break;
            case '5':
                save();
                console.log('Goodbye!');
                rl.close();
                return;
            default:
                console.log('Invalid choice.');
        }
    }
}

main().catch(console.error);

File: budget_tracker.go
// budget_tracker.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

const dataFile = "budget.json"

type Transaction struct {
	Type     string  `json:"type"`
	Category string  `json:"category"`
	Amount   float64 `json:"amount"`
	Date     string  `json:"date"`
}

var transactions []Transaction

func load() {
	file, err := os.ReadFile(dataFile)
	if err != nil {
		transactions = []Transaction{}
		return
	}
	if err := json.Unmarshal(file, &transactions); err != nil {
		transactions = []Transaction{}
	}
}

func save() error {
	data, err := json.MarshalIndent(transactions, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(dataFile, data, 0644)
}

func addTransaction(ttype, category string, amount float64, dateStr string) bool {
	if dateStr == "" {
		dateStr = time.Now().Format("2006-01-02")
	} else {
		// validate format
		_, err := time.Parse("2006-01-02", dateStr)
		if err != nil {
			fmt.Println("Invalid date format. Use YYYY-MM-DD.")
			return false
		}
	}
	transactions = append(transactions, Transaction{
		Type:     ttype,
		Category: category,
		Amount:   amount,
		Date:     dateStr,
	})
	if err := save(); err != nil {
		fmt.Println("Error saving:", err)
		return false
	}
	return true
}

func getBalance() (income, expense, balance float64) {
	for _, t := range transactions {
		if t.Type == "income" {
			income += t.Amount
		} else if t.Type == "expense" {
			expense += t.Amount
		}
	}
	balance = income - expense
	return
}

func categoryStats(ttype string) map[string]float64 {
	stats := make(map[string]float64)
	for _, t := range transactions {
		if t.Type == ttype {
			stats[t.Category] += t.Amount
		}
	}
	return stats
}

func printChart(ttype string, width int) {
	stats := categoryStats(ttype)
	if len(stats) == 0 {
		fmt.Println("No data to chart.")
		return
	}
	var total float64
	maxVal := 0.0
	for _, v := range stats {
		total += v
		if v > maxVal {
			maxVal = v
		}
	}
	fmt.Printf("\n%s chart (by category):\n", strings.Title(ttype))
	for cat, val := range stats {
		barLen := int((val / maxVal) * float64(width))
		bar := strings.Repeat("█", barLen)
		fmt.Printf("%-12s %s %.2f\n", cat, bar, val)
	}
	fmt.Printf("Total: %.2f\n", total)
}

func main() {
	load()
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("=== Budget Tracker ===")
	for {
		fmt.Println("\n1. Add transaction")
		fmt.Println("2. Show balance")
		fmt.Println("3. Show category stats")
		fmt.Println("4. Show expense chart")
		fmt.Println("5. Save & exit")
		fmt.Print("Choose: ")
		scanner.Scan()
		choice := strings.TrimSpace(scanner.Text())
		switch choice {
		case "1":
			fmt.Print("Type (income/expense): ")
			scanner.Scan()
			ttype := strings.ToLower(strings.TrimSpace(scanner.Text()))
			if ttype != "income" && ttype != "expense" {
				fmt.Println("Invalid type.")
				continue
			}
			fmt.Print("Category: ")
			scanner.Scan()
			category := strings.TrimSpace(scanner.Text())
			fmt.Print("Amount: ")
			scanner.Scan()
			amount, err := strconv.ParseFloat(strings.TrimSpace(scanner.Text()), 64)
			if err != nil {
				fmt.Println("Invalid amount.")
				continue
			}
			fmt.Print("Date (YYYY-MM-DD, leave blank for today): ")
			scanner.Scan()
			dateStr := strings.TrimSpace(scanner.Text())
			if addTransaction(ttype, category, amount, dateStr) {
				fmt.Println("Transaction added.")
			}
		case "2":
			inc, exp, bal := getBalance()
			fmt.Printf("\nIncome: %.2f\n", inc)
			fmt.Printf("Expenses: %.2f\n", exp)
			fmt.Printf("Balance: %.2f\n", bal)
		case "3":
			stats := categoryStats("expense")
			if len(stats) == 0 {
				fmt.Println("No expenses recorded.")
			} else {
				var total float64
				for _, v := range stats {
					total += v
				}
				fmt.Println("\nExpense categories:")
				for cat, val := range stats {
					pct := (val / total * 100)
					fmt.Printf("  %-12s %10.2f  (%5.1f%%)\n", cat, val, pct)
				}
			}
		case "4":
			printChart("expense", 40)
		case "5":
			if err := save(); err != nil {
				fmt.Println("Error saving:", err)
			} else {
				fmt.Println("Goodbye!")
			}
			return
		default:
			fmt.Println("Invalid choice.")
		}
	}
}
